### PR TITLE
Ensure that we do not re-use CompilationWithAnalyzers in presence of cancellation

### PR DESCRIPTION
**Pull request by mavasani**
Recommended to review each commit separately:
1. https://github.com/dotnet/roslyn/commit/1e75ba71ac55e9270feb4ff67524c9f48f7e6601: Handles in-proc behavior change, with the added unit test that fails prior to product changes in the commit
2. https://github.com/dotnet/roslyn/commit/505826d1620d2549500761bafa62f649fa427350: Handles out-of-proc behavior change, with the added unit test that fails prior to product changes in the commit

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 1]
```diff
@@ -147,9 +150,17 @@ public async Task<IEnumerable<DiagnosticData>> ComputeDiagnosticsAsync(Diagnosti
...
-            var resultAndTelemetry = await _diagnosticAnalyzerRunner.AnalyzeDocumentAsync(analysisScope, _compilationWithAnalyzers,
-                _logPerformanceInfo, getTelemetryInfo: false, cancellationToken).ConfigureAwait(false);
-            return resultAndTelemetry.AnalysisResult;
+            try
+            {
+                var resultAndTelemetry = await _diagnosticAnalyzerRunner.AnalyzeDocumentAsync(analysisScope, _compilationWithAnalyzers,
+                    _logPerformanceInfo, getTelemetryInfo: false, cancellationToken).ConfigureAwait(false);
+                return resultAndTelemetry.AnalysisResult;
+            }
+            catch (OperationCanceledException)
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]: ❔ Why do we want to execute this callback on cancellation, but _not_ execute the callback for some other exception?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 2]
```diff
@@ -77,7 +77,7 @@ private async Task AnalyzeDocumentForKindAsync(TextDocument document, AnalysisKi
                 if (nonCachedStateSets.Count > 0)
                 {
                     var analysisScope = new DocumentAnalysisScope(document, span: null, nonCachedStateSets.SelectAsArray(s => s.Analyzer), kind);
-                    var executor = new DocumentAnalysisExecutor(analysisScope, compilationWithAnalyzers, _diagnosticAnalyzerRunner, logPerformanceInfo: true);
+                    var executor = new DocumentAnalysisExecutor(analysisScope, compilationWithAnalyzers, _diagnosticAnalyzerRunner, logPerformanceInfo: true, onAnalysisCanceled: OnAnalysisCanceled);
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:80]: ❔ Can we do this without adding more per-call closure allocations?

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]: That was the only known case where we are seeing failures. I am fine with making this more generic, though I would not be able to add relevant tests.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:80]: Sure.

Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:80]: Actually, the `DocumentAnalysisExecutor` is created just once in this method, so it should already be a single allocation?

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]: I would prefer to take the more conservative approach here (don't reuse on cancellation or failure). Tests for the cancellation case should be sufficient.

Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]: Sounds good to me, let me adjust.

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 3]
```diff
@@ -156,9 +156,9 @@ public async Task<IEnumerable<DiagnosticData>> ComputeDiagnosticsAsync(Diagnosti
                     _logPerformanceInfo, getTelemetryInfo: false, cancellationToken).ConfigureAwait(false);
                 return resultAndTelemetry.AnalysisResult;
             }
-            catch (OperationCanceledException)
+            catch (Exception)
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]:
```suggestion
            catch
```

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysis/Diagnostics/DiagnosticComputer.cs [hunk 4]
```diff
@@ -85,8 +85,9 @@ internal class DiagnosticComputer
                 return await AnalyzeAsync(compilationWithAnalyzers, analyzerToIdMap, analyzers, skippedAnalyzersInfo,
                     reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken).ConfigureAwait(false);
             }
-            catch (OperationCanceledException)
+            catch (Exception)
```
Comment (sharwell) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysis/Diagnostics/DiagnosticComputer.cs:88]:
```suggestion
            catch
```

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs (hunk 3 above)
Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:159]: Sure, a code style analyzer that fades it out would be great :-)

**Pull request merged**